File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-3.js
class Solution {
    /**
     * @param {number[]} cost
     * @return {number}
     */
    minCostClimbingStairs(cost) {
        const cache = new Map()

        const clacCostFrom = (i) => {
            if (i >= cost.length) return 0
            if (cache.has(i)) return cache.get(i)
            const v = Math.min(clacCostFrom(i + 1), clacCostFrom(i + 2)) + cost[i]
            cache.set(i, v)
            return v
        }

        const res = Math.min(clacCostFrom(0), clacCostFrom(1))
        return res
    }
}

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-3.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l = 0
        r = len(nums) - 1
        while r - l > 1:
            m = (l + r) // 2
            if nums[l] <= nums[m]:
                l = m
            else:
                r = m
        return min(nums[0], nums[r])